tests/Feature/Http/Controllers/Api/V2/ApplicationApiControllerTest.php

```diff
@@ -154,4 +155,90 @@ public function it_allows_null_device_id()
         $response->assertNotFound();
     }
 
+    /**
+     * @test
+     */
+    public function it_return_the_update_when_device_id_is_in_partial_rollout_range(): void
+    {
+        // Create the application
+        $application = $this->makeApplicationModel();
+
+        // Select the platform
+        $platform = new AndroidPlatform();
+
+        $build = $this->makeBuildModel($application->id, $platform->getId(), [
+            'version' => '1.0.0',
+            'available_from' => Carbon::today()->subWeek(),
+            'partial_rollout' => true,
+            'rollout_percentage' => 50,
+        ]);
+
+        $devices = factory(Device::class, 20)->create([
+            'build_id' => $build->id,
+            'application_id' => $application->id,
+        ]);
+
+        $updateAvailable = 0;
+        foreach ($devices as $device) {
+            $response = $this->post(route('api.v2.updates.get', [
+                'application' => $application->slug,
+                'platform' => $platform->getId(),
+            ]), [
+                'version' => '0.9.9',
+                'device_id' => $device->device_id
+            ]);
+
+            if ($response->json('data') === null) {
+                $updateAvailable++;
+            }
+        }
+
+        $this->assertSame(10, $updateAvailable);
+    }
+
+    /**
+     * @test
+     */
+    public function it_return_the_last_build_if_device_is_not_in_the_partial_rollout_range(): void
```

Comment:
Can you please add a test without `$build1`. I'm worried that that `while...do` may potentially cause an infinite loop

Reply:
Added another test, the issue was not the infinite loop but a corner case when the rollout percentage is zero (which doesn't make much sense but still a broken behaviour) which will return that the device is always in the range.  Fixed in https://github.com/everli/uppy/pull/8/commits/853108fb3be394d4edaf40d79277865e1bce67f9